cmd/launcher/extension.go

```diff
@@ -0,0 +1,160 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"crypto/x509"
+	"io/ioutil"
+
+	"github.com/boltdb/bolt"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	kolidelog "github.com/kolide/launcher/pkg/log"
+	"github.com/kolide/launcher/pkg/osquery"
+	"github.com/kolide/launcher/pkg/osquery/runtime"
+	"github.com/kolide/launcher/pkg/osquery/table"
+	"github.com/kolide/launcher/pkg/service"
+	"github.com/kolide/osquery-go/plugin/config"
+	"github.com/kolide/osquery-go/plugin/distributed"
+	osquerylogger "github.com/kolide/osquery-go/plugin/logger"
+	"github.com/pkg/errors"
+)
+
+// extension is the osquery extension and osquery runtime
+type extension struct {
+	*runtime.Runner
+	Execute   func() error
+	Interrupt func(err error)
+}
+
+// TODO: the extension, runtime, and client are all kind of entangled here. Untangle the underlying libraries and separate into units
+func createExtension(ctx context.Context, rootDirectory string, db *bolt.DB, logger log.Logger, opts *options) (*extension, error) {
+	// read the enroll secret, if either it or the path has been specified
+	var enrollSecret string
+	if opts.enrollSecret != "" {
+		enrollSecret = opts.enrollSecret
+	} else if opts.enrollSecretPath != "" {
+		content, err := ioutil.ReadFile(opts.enrollSecretPath)
+		if err != nil {
+			return nil, errors.Wrapf(err, "could not read enroll_secret_path: %s", opts.enrollSecretPath)
+		}
+		enrollSecret = string(bytes.TrimSpace(content))
+	}
+
+	// create the certificate pool
+	var rootPool *x509.CertPool
+	if opts.rootPEM != "" {
+		rootPool = x509.NewCertPool()
+		pemContents, err := ioutil.ReadFile(opts.rootPEM)
+		if err != nil {
+			return nil, errors.Wrapf(err, "reading root certs PEM at path: %s", opts.rootPEM)
+		}
+		if ok := rootPool.AppendCertsFromPEM(pemContents); !ok {
+			return nil, errors.Errorf("found no valid certs in PEM at path: %s", opts.rootPEM)
+		}
+	}
+
+	// connect to the grpc server
+	grpcConn, err := service.DialGRPC(opts.kolideServerURL, opts.insecureTLS, opts.insecureGRPC, opts.certPins, rootPool, logger)
+	if err != nil {
+		return nil, errors.Wrap(err, "dialing grpc server")
+	}
+
+	// create the client of the grpc service
+	launcherClient := service.New(grpcConn, level.Debug(logger))
+
+	// create the osquery extension
+	extOpts := osquery.ExtensionOpts{
+		EnrollSecret:    enrollSecret,
+		Logger:          logger,
+		LoggingInterval: opts.loggingInterval,
+	}
+
+	// create the extension
+	ext, err := osquery.NewExtension(launcherClient, db, extOpts)
+	if err != nil {
+		return nil, errors.Wrap(err, "starting grpc extension")
+	}
+
+	// create the logging adapter for osquery
+	osqueryLogger := &kolidelog.OsqueryLogAdapter{Logger: level.Debug(log.With(logger, "component", "osquery"))}
+
+	var runner *runtime.Runner
```

Comment:
can you add a comment explaining why a nil runner is added to the struct, rather than creating the runner here? 

I understand after reading the code, but it would help with readability.